docs/development/design/github-application.rst

```diff
@@ -0,0 +1,306 @@
+``readthedocs/blog``" should "trigger a build for ``latest`` version on
+``read-the-docs-blog`` project". For these cases we can use use
+``repository.id`` field from the body to get the ``RemoteRepository.remote_id``
+and from there get the ``Project``.
+
+
+Remote* models re-sync
+----------------------
+
+Currently, we are using 2 endpoints to sync all the ``Remote*`` models:
+
+* ``https://api.github.com/user/repos`` (https://docs.github.com/en/rest/reference/repos#list-repositories-for-the-authenticated-user)
+* ``https://api.github.com/org/{org}/repos`` (https://docs.github.com/en/rest/reference/repos#list-organization-repositories)
+
+However, this endpoints won't return the same data when using GitHub Application
+since we won't be authenticated as a user anymore and the ``permission.admin:
+boolean`` field won't come in the response.
+
+Instead, we will have to iterate over,
+
+* all repositories accessible to the app installation
+
+  * ``https://api.github.com/installation/repositories`` (https://docs.github.com/en/rest/reference/apps#list-repositories-accessible-to-the-app-installation)
+* iterate over each repository checking for user's permission
+
+  * ``https://api.github.com/repos/{owner}/{repo}/collaborators/{username}/permission`` (https://docs.github.com/en/rest/reference/repos#get-repository-permissions-for-a-user)
+
+By doing this, we will keep our ``Remote*`` table very small because we will
+only track repositories that users gave us permissions. Then, only these
+repositories will be shown in the "Import Project" page.
+
+Once they created a new repository under their organization, they will need to
+go to the GitHub Application installation configuration and grant access to the
+new repository. This will send us a webhook (``installation_repositories``,
+https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#installation_repositories)
+and we can automatically do a re-sync of ``Remote*`` models.
+
+.. note::
+
+   Users installing our GitHub Application and selecting "All repositories" will
+   always see all the repositories on the "Import Project" page.
+
+
+Need more research
+------------------
+
+* Does django-allauth support GitHub Application for Login?
+
+  Using the Client ID for the GitHub Application (instead the OAuth
+  application), should make it to work:
+  https://docs.github.com/en/developers/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps#web-application-flow
+
+* How do we use Client ID for GitHub Application for new users and Client ID for
+  OAuth application for current/existing users?
+
+* Should we keep using our OAuth Application for login and GitHub Application
+  for the rest? Is it possible?
+
+  Related: https://docs.github.com/en/developers/apps/getting-started-with-apps/migrating-oauth-apps-to-github-apps
+  Related: https://github.com/readthedocs/readthedocs-ops/issues/532
+
```

Comment:
These seem like really important questions. The interaction with allauth in particular worries me. I really don't want to have to hack a lot of stuff, so keeping auth & repo permissions separated might be the best solution. Could even make the GH App approach optional for more security-minded people?

Reply:
> so keeping auth & repo permissions separated might be the best solution

I'm afraid that this is not possible. At the "Login" page, when the user clicks on "Login with GitHub" we need to redirect them to the URL to authorize the OAuth Application or the GitHub Application --so we have to make that decision before showing the button. I don't really know how to solve this problem yet. This will require a little more thinking and research.

> Could even make the GH App approach optional for more security-minded people?

This question makes me think that we may need to show 2 different buttons: OAuth App and GH App. I know that would be terrible UX and will confuse a lot of users, tho.

I'd say that we should keep OAuth working as long as we have old users using this integration. However, *new users should not have the option to use OAuth App* and they must use GH App for all the good reasons we are migrating to it.

At some point, when no users using OAuth left, we should be able to delete it completely.

The most important thing to solve here is "How do we know that the person on the Login page is old or new to show them the appropriate button?"

Reply:
Agreed having two buttons would be bad UX, we'd confuse users a lot here.

Is there any way we can try both, starting with one and if that attempt fails, try the other automatically? Or do either allow us to query to see if it is installed for the user?

To start, we could attempt authentication with the oauth app, and back down to the GH app -- swapping once we feel more users are using the GH app.

This could leave a single button in tact but for some users there would be an ugly wait to log in as we try to auth twice

> new users should not have the option to use OAuth App and they must use GH App

:+1:

Reply:
Honestly, I'm not finding a good way to solve this problem. I re-read the migration guide that GitHub provides and it's not super useful in this case.

I was thinking of a different approach:

1. implement all the GH application code
2. do not expose GH application to users yet, but make all the URLs to work
3. QA internally and make sure that everything works as expected
4. send emails + add a banner (with "call to action") telling organization owners to install our GH application (https://docs.github.com/en/developers/apps/getting-started-with-apps/migrating-oauth-apps-to-github-apps#direct-users-to-install-your-github-app-on-repositories)
5. send another email + add a banner communicating the EOL of the OAuth and telling them that starting on X date they must have GH App installed to use Read the Docs

As always, dropping support for something at a specific date is not good; but I'm not clearly seeing what else we can do besides forcing users to migrate. This may be doable in commercial, but probably won't in community.

Anyways, I'm open to getting some feedback and ideas on this because I ran out of ideas 😞

Reply:
You might have the most knowledge and best opinions here. I'm not super familiar with the implementation at a technical level so can only take some wild shots in the dark.

The thing that I find awkward about that approach is that the user will get this notification to use the GHA method, allow the application on their user/organization, but then they need both the GHA and the oauth app installed until we do the cut over to the GHA.

It feels like we want users to have a different login entirely if they want to use the GHA implementation. Does it make sense to discuss using the beta instance for GHA in addition to the new templates?